blockchain/pool.go

```diff
@@ -1,630 +1,372 @@
-			requester.redo(peerID)
+// Stops the peer timer and deletes the peer. Recomputes the max peer height.
+func (pool *blockPool) deletePeer(peerID p2p.ID) {
+	if p, ok := pool.peers[peerID]; ok {
+		if p.timeout != nil {
+			p.timeout.Stop()
 		}
-	}
+		delete(pool.peers, peerID)
 
-	peer, ok := pool.peers[peerID]
-	if ok {
-		if peer.timeout != nil {
-			peer.timeout.Stop()
+		if p.height == pool.maxPeerHeight {
+			pool.updateMaxPeerHeight()
 		}
+	}
+}
 
-		delete(pool.peers, peerID)
+// Removes any blocks and requests associated with the peer and deletes the peer.
+// Also triggers new requests if blocks have been removed.
+func (pool *blockPool) removePeer(peerID p2p.ID, err error) {
+	pool.logger.Debug("removing peer", "peerID", peerID)
 
-		// Find a new peer with the biggest height and update maxPeerHeight if the
-		// peer's height was the biggest.
-		if peer.height == pool.maxPeerHeight {
-			pool.updateMaxPeerHeight()
+	peer := pool.peers[peerID]
+	if peer == nil {
+		return
+	}
+	// Reschedule the requests for all blocks waiting for the peer, or received and not processed yet.
+	for h, block := range pool.peers[peerID].blocks {
+		if block == nil {
+			pool.numPending--
 		}
+		pool.rescheduleRequest(peerID, h)
 	}
+	pool.deletePeer(peerID)
+
 }
 
-// If no peers are left, maxPeerHeight is set to 0.
-func (pool *BlockPool) updateMaxPeerHeight() {
-	var max int64
+// Called every time FSM advances its height.
+func (pool *blockPool) removeShortPeers() {
 	for _, peer := range pool.peers {
-		if peer.height > max {
-			max = peer.height
+		if peer.height < pool.height {
+			pool.removePeer(peer.id, nil)
 		}
 	}
-	pool.maxPeerHeight = max
 }
 
-// Pick an available peer with at least the given minHeight.
-// If no peers are available, returns nil.
-func (pool *BlockPool) pickIncrAvailablePeer(minHeight int64) *bpPeer {
-	pool.mtx.Lock()
-	defer pool.mtx.Unlock()
-
+func (pool *blockPool) removeBadPeers() {
+	pool.removeShortPeers()
 	for _, peer := range pool.peers {
-		if peer.didTimeout {
-			pool.removePeer(peer.id)
-			continue
-		}
-		if peer.numPending >= maxPendingRequestsPerPeer {
-			continue
-		}
-		if peer.height < minHeight {
-			continue
+		if err := peer.isGood(); err != nil {
+			pool.removePeer(peer.id, err)
+			if err == errSlowPeer {
+				peer.errFunc(errSlowPeer, peer.id)
+			}
 		}
-		peer.incrPending()
-		return peer
 	}
-	return nil
 }
 
-func (pool *BlockPool) makeNextRequester() {
-	pool.mtx.Lock()
-	defer pool.mtx.Unlock()
-
-	nextHeight := pool.height + pool.requestersLen()
-	if nextHeight > pool.maxPeerHeight {
-		return
+func (pool *blockPool) makeRequestBatch(maxNumPendingRequests int32) []int {
+	pool.removeBadPeers()
+	// If running low on planned requests, make more.
+	numNeeded := int32(cmn.MinInt(int(maxNumPendingRequests), len(pool.peers)*int(maxRequestsPerPeer))) - pool.numPending
+	for int32(len(pool.requests)) < numNeeded {
+		if pool.nextRequestHeight > pool.maxPeerHeight {
```

Comment:
Do you handle a case where a request failed and you can find a new peer that can serve that request at that point of time, but some time after there is a peer that can provide it? Imagine for example that there is a peer p that can serve failed request but we can't send to him as number of pending peer requests for p is reached. The problematic case is the one where there is a hole in a sequence. In that case you want to send request for such a block as soon as possible, and this can be lower than `pool.nextRequestHeight`.

Reply:
If a request cannot be sent it is left in the requests[] list and retried later. The list is sorted by height so the lowest block in the list will be the one requested first.